specs/emip-004.md

```diff
@@ -0,0 +1,67 @@
+
+# Motivation
+
+We need a specification to describe a general-purpose scheme for generating private key chains which can be adapted for specific functionalities. For example, the memo feature in Yoroi needs to sign and encrypt memo data with different keys without asking the user for a password. To do so, we look for a method to derive new keys from the existing wallet's account keys but also in the case of the wallet providing only the public key of the account, like in hardware wallets, or even when the wallet does not provide keys at all (public or private), like in the case of *balance tracker* wallets.
+
+# Background
+
+[BIP32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#Specification_Key_derivation) introduced the concept of a hierarchical deterministic (HD) wallet, a logical hierarchy structure built using child key derivation functions (CKD) that allows to generate multiple wallets from a master seed. 
+
+![image](https://raw.githubusercontent.com/bitcoin/bips/master/bip-0032/derivation.png)
+
+[BIP43](https://github.com/bitcoin/bips/blob/master/bip-0043.mediawiki) added a purpose field for derivations (derivation path) and [BIP44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki) introduced the concept of multi-account hierarchy deterministic wallets, defining 5 leves of derivation.
+
+`m / purpose' / coin_type' / account' / change / address_index`
+
+You can find a list of registered purposes [here](https://wiki.trezor.io/Cryptocurrency_standards). Yoroi is an example of a HD wallet following the BIP44 standard. Yoroi addresses use the following derivation path:
+
+`m / 44' / 1815' / 0' / 0 / 1`
+
+In this example derivation, the purpose is 44 (multi-account HD wallet), the coin type is 1815 (Cardano), the account index is 0 (first account), the change index is also 0, and the address index is 1 (second address in the account).
+
+# Master seed and Private chain
+
+There are three possibilites when considering the availability of the wallet account master key pair (which from now on we call the *account key pair*):
+
+* **Standard Yoroi wallets**: Both the public and the private keys of the account key pair are available.
+* **Hardware wallets**: Only the public key is available.
+* **Balance tracker wallets**: No keys are available.
+
+Based on the above, we identify two scenarios:
+
+* **S1**: account public key is available
+* **S2**: account public key is not available
+
+In both cases, even if the account public key is available, we don't derive the child keys directly from the account as we don't want to conflict with any other derivations. We discuss each scenario in the following sections. 
+
+## Public key is available
+
+We create a new key pair from the account public key in the following way:
+
+* Hash the account public key with `Blake2b` and specify an output length of 32 bytes.
```

Comment:
Since, wallet users should know that they shouldn't reveal account public key for privacy reasons, the only bad thing I see here is a predictable entropy injection - when user (by modifying wallet configuration) is able to replace account public key on some low-entropy value, while memo chain generation algorithm executing. Considering this, probably, it would be useful to store account public key encrypted (for example using wallet password)